src/txadmission.cpp

```diff
@@ -0,0 +1,1028 @@
+}
+
+void ThreadCommitToMempool()
+{
+    while (!ShutdownRequested())
+    {
+        {
+            boost::unique_lock<boost::mutex> lock(csCommitQ);
+            do
+            {
+                cvCommitQ.timed_wait(lock, boost::posix_time::milliseconds(2000));
+            } while (txCommitQ.empty() && txDeferQ.empty());
+        }
+
+        {
+            boost::this_thread::interruption_point();
+
+            CORRAL(txProcessingCorral, CORRAL_TX_COMMITMENT);
+            LOCK(cs_main);
+            CommitTxToMempool();
+            mempool.check(pcoinsTip);
+            LOG(MEMPOOL, "MemoryPool sz %u txn, %u kB\n", mempool.size(), mempool.DynamicMemoryUsage() / 1000);
+            // BU - Xtreme Thinblocks - trim the orphan pool by entry time and do not allow it to be overidden.
+            LimitMempoolSize(mempool, GetArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE) * 1000000,
+                GetArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY) * 60 * 60);
+
+            // BU: update tx per second when a tx is valid and accepted
+            mempool.UpdateTransactionsPerSecond();
+            CValidationState state;
+            FlushStateToDisk(state, FLUSH_STATE_PERIODIC);
+            // The flush to disk above is only periodic therefore we need to continuously trim any excess from the
+            // cache.
+            pcoinsTip->Trim(nCoinCacheMaxSize);
+
+            // move the previously deferred txs into active processing
+            if (!txDeferQ.empty())
+            {
+                // this could be a lot more efficient
+                std::queue<CTxInputData> wasDeferred;
+                while (!txDeferQ.empty())
+                {
+                    wasDeferred.push(txDeferQ.front());
+                    txDeferQ.pop();
+                }
+                LOG(MEMPOOL, "%d tx were deferred\n", wasDeferred.size());
+
+                while (!wasDeferred.empty())
+                {
+                    //LOG(MEMPOOL, "attempt enqueue deferred %s\n", wasDeferred.front().tx->GetHash().ToString());
+                    EnqueueTxForAdmission(wasDeferred.front());
+                    wasDeferred.pop();
+                }
+            }
+        }
+    }
+}
+
+void LimitMempoolSize(CTxMemPool &pool, size_t limit, unsigned long age)
+{
+    std::vector<COutPoint> vCoinsToUncache;
+    int expired = pool.Expire(GetTime() - age, vCoinsToUncache);
+    for (const COutPoint &txin : vCoinsToUncache)
+        pcoinsTip->Uncache(txin);
+    if (expired != 0)
+        LOG(MEMPOOL, "Expired %i transactions from the memory pool\n", expired);
+
+    std::vector<COutPoint> vNoSpendsRemaining;
+    pool.TrimToSize(limit, &vNoSpendsRemaining);
+    for (const COutPoint &removed : vNoSpendsRemaining)
+        pcoinsTip->Uncache(removed);
+}
+
+void CommitTxToMempool()
+{
+    std::vector<uint256> whatChanged;
+    LOCK(cs_main); // cs_main must lock before csCommitQ
+    {
```

Comment:
why a lock on cs_main, isnt  it txCommitQ that needs a lock...or do we possibly need both here ?  LOCK2(cs_main , cstxCommitQ); ?

Reply:
oh , duh I didn't see the lock on csCommitQ, I was expecting to see WRITELOCK...ok that looks good